crates/libs/core/src/sync/cancel.rs

```diff
@@ -0,0 +1,729 @@
+    //     Ok(self.rx.blocking_recv().unwrap())
+    // }
+
+    // Set the SF ctx to hook up cancellation.
+    fn set_ctx(&mut self, ctx: IFabricAsyncOperationContext) {
+        let prev = self.ctx.replace(ctx);
+        assert!(prev.is_none());
+    }
+
+    // Cancels the inner SF operation if exists, and reset the ctx.
+    fn cancel_inner_ctx(&mut self) -> crate::Result<()> {
+        if let Some(ctx) = &self.ctx {
+            if let Err(e) = unsafe { ctx.Cancel() } {
+                // fail to cancel inner operation.
+                return Err(e);
+            } else {
+                // clear the sf ctx to avoid cancel twice.
+                self.ctx.take();
+            }
+        }
+        Ok(())
+    }
+}
+
+// Returns error if cancelled.
+// If there is an inner SF ctx, cancellation signal will
+// trigger cancellation of the ctx.
+impl<T> Future for FabricReceiver2<T> {
+    // The error code should be OperationCanceled, unless cancellation
+    // of SF ctx returns other errors.
+    // (TODO: observe other error code from SF, maybe some code should be ignored).
+    type Output = crate::Result<T>;
+    fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Poll the receiver first, if ready then return the output,
+        // else poll the cancellation token, if cancelled return the error,
+        // else return pending.
+        // There can be the case that cancellation wakes the waker, but receiver
+        // then got the result. The next poll will return received output rather
+        // than cancelled error.
+
+        // Try to receive the value from the sender
+        let innner =
+            <tokio::sync::oneshot::Receiver<T> as Future>::poll(Pin::new(&mut self.rx), _cx);
+        match innner {
+            Poll::Ready(x) => {
+                // error only happens when sender is dropped without sending.
+                match x {
+                    Ok(data) => Poll::Ready(Ok(data)),
+                    Err(_) => {
+                        if let Some(t) = self.token.as_ref() {
+                            if t.is_cancelled() {
+                                // cancel the SF ctx
+                                if let Err(e) = self.cancel_inner_ctx() {
+                                    return Poll::Ready(Err(e));
+                                }
+                                return Poll::Ready(Err(FabricErrorCode::OperationCanceled.into()));
+                            }
+                        }
+                        panic!("sender dropped without sending")
+                    }
+                }
+            }
+            Poll::Pending => {
+                // If the action is canceled we can safely stop and return canceled error.
+                if let Some(t) = &self.token {
+                    // this is cancel safe so we can poll it once and discard
+                    let fu = t.cancelled();
```

Comment:
I think you can flatten this by a couple levels of indentation:
```suggestion
        match (inner, self.token) {
            (Poll::Ready(Ok(data)), _) => {
                 Poll::Ready(Ok(data))
             },
             (Poll::Ready(Err(_)),  Some(t)) => {
                 if t.is_cancelled() {
                       // cancel the SF ctx
                      if let Err(e) = self.cancel_inner_ctx() {
                         return Poll::Ready(Err(e));
                      }
                     return Poll::Ready(Err(FabricErrorCode::OperationCanceled.into()));
                }
            }
            (Poll::Pending, Some(t) => {
                // If the action is canceled we can safely stop and return canceled error.
                // this is cancel safe so we can poll it once and discard
               let fu = t.cancelled();
```

Reply:
Flattened. It has less indentations but more cases.